go/vt/vttablet/tabletserver/txlimiter/tx_limiter.go

```diff
@@ -0,0 +1,207 @@
+	vtrpcpb "github.com/youtube/vitess/go/vt/proto/vtrpc"
+)
+
+const unknown string = "unknown"
+
+var (
+	rejections       = stats.NewCounters("TxLimiterRejections")
+	rejectionsDryRun = stats.NewCounters("TxLimiterRejectionsDryRun")
+)
+
+// TxLimiter is the transaction limiter interface.
+type TxLimiter interface {
+	Get(ctx context.Context) bool
+	ReleaseByContext(ctx context.Context)
+	Release(immediate *querypb.VTGateCallerID, effective *vtrpcpb.CallerID)
+}
+
+// New creates a new TxLimiter.
+// slotCount: total slot count in transaction pool
+// maxPerUser: fraction of the pool that may be taken by single user
+// enabled: should the feature be enabled. If false, will return
+// "allow-all" limiter
+// dryRun: if true, does no limiting, but records stats of the decisions made
+// byXXX: whether given field from immediate/effective caller id should be taken
+// into account when deciding "user" identity for purposes of transaction
+// limiting.
+func New(slotCount int, maxPerUser float64, enabled, dryRun, byUsername, byPrincipal, byComponent, bySubcomponent bool) TxLimiter {
+	if !enabled && !dryRun {
+		return &TxAllowAll{}
+	}
+
+	return &Impl{
+		maxPerUser:      int64(float64(slotCount) * maxPerUser),
+		dryRun:          dryRun,
+		byUsername:      byUsername,
+		byPrincipal:     byPrincipal,
+		byComponent:     byComponent,
+		bySubcomponent:  bySubcomponent,
+		byEffectiveUser: byPrincipal || byComponent || bySubcomponent,
+		usageMap:        make(map[string]int64),
+	}
+}
+
+// TxAllowAll is a TxLimiter that allows all Get requests and does no tracking.
+// Implements Txlimiter.
+type TxAllowAll struct{}
+
+// Get always returns true (allows all requests).
+// Implements TxLimiter.Get
+func (txa *TxAllowAll) Get(ctx context.Context) bool {
+	return true
+}
+
+// ReleaseByContext is noop, because TxAllowAll does no tracking.
+// Implements TxLimiter.ReleaseByContext
+func (txa *TxAllowAll) ReleaseByContext(ctx context.Context) {
+
+}
+
+// Release is noop, because TxAllowAll does no tracking.
+// Implements TxLimiter.Release
+func (txa *TxAllowAll) Release(immediate *querypb.VTGateCallerID, effective *vtrpcpb.CallerID) {
+	// NOOP
+}
+
+// Impl limits the total number of transactions a single user may use
+// concurrently.
+// Implements TxLimiter.
+type Impl struct {
+	maxPerUser int64
+	usageMap   map[string]int64
+	mu         sync.Mutex
+
+	dryRun          bool
+	byUsername      bool
+	byPrincipal     bool
+	byComponent     bool
+	bySubcomponent  bool
+	byEffectiveUser bool
+}
+
+// Get tells whether given user (identified by context.Context) is allowed
+// to use another transaction slot. If this method returns true, it's
+// necessary to call Release once transaction is returned to the pool.
+// Implements TxLimiter.Get
+func (txl *Impl) Get(ctx context.Context) bool {
+	key := txl.extractKeyFromContext(ctx)
+
+	txl.mu.Lock()
+	defer txl.mu.Unlock()
+
+	usage, ok := txl.usageMap[key]
+	if !ok {
```

Comment:
Just `usage := txl.usageMap[key]` should be enough. If absent, usage will be 0, and will naturally flow into the next check.

Reply:
Right. Done.